### DISCUSSION TITLE
How to wait all packet are sent before StreamShutdown/Close

### DISCUSSION THREAD

**Discussion by ami-GS** [microsoft/msquic#2187] [Q&A] [answered]
API doc said that StreamSend just enqueue data on stream and it is unblocking call.
How can the app wait the enqueued data to be sent before StreamShutdown/Close?

**Comment by ThadHouse** [comment, marked as answer]
You can call shutdown while there is still data in flight, as long as it's not an abortive shutdown. that will just shut down the send side, you can still receive data. Then you can wait for the ShutdownComplete callback, and close once you receive that.

**Comment by ami-GS** [comment]
My code example is bellow.
I experimented a bit. You are right, the data was sent as expected by putting sleep **before** `MsQuic->ConnectionShutdown()`
This means the data on stream was sent between `MsQuic->StreamShutdown()` and QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE.
Problem should be come from `ConnectionShutdown/Close` before QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE on the stream.
Should `ConnectionShutdown` called after QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE?

```c
void close(Stream, Connection, Configuration, Registration) {
    if (Stream) {
        MsQuic->StreamShutdown(Stream, QUIC_STREAM_SHUTDOWN_FLAG_GRACEFUL, 0);
    }

    //sleep(1);
    if (Connection) {
        MsQuic->ConnectionShutdown(Connection, QUIC_CONNECTION_SHUTDOWN_FLAG_NONE, 0);
    }

    if (Configuration) {
        MsQuic->ConfigurationClose(Configuration);
    }

    if (Registration) {
        MsQuic->RegistrationShutdown(Registration, QUIC_CONNECTION_SHUTDOWN_FLAG_NONE, 0);
    }
}
```

**Reply by ThadHouse** [reply]
Yes, you have to wait for the stream shutdown event to complete before calling ConnectionShutdown. ConnectionShutdown will implicitly close all streams. See the remarks docs for connection shutdown (https://github.com/microsoft/msquic/blob/main/docs/api/ConnectionShutdown.md) for a bit more info in this.
